app/src/main/java/org/sil/gatherwords/SessionActivity.java

```diff
@@ -118,18 +129,127 @@ public void save_settings_fab_pressed(View view) {
         startActivity(i);
     }
 
-    private static class InsertSessionsTask extends AsyncTask<Session, Void, Void> {
+    private static class InsertSessionsTask extends AsyncTask<Session, Double, Void> {
         private SessionDao sDAO;
+        private WordDao wordDao;
+        // TODO: fix null object reference so we don't risk a data leak with the context
+//        private final ThreadLocal<SessionActivity> sessionActivity = new ThreadLocal<>();
+        private Long sessionID;
+        private int maxProgress;
+        private AppDatabase db;
+        private String wordList;
+        private final Context context;
```

Comment:
I see.  It is `WeakReference` you want, not `final` (you could use that if you want access to the entire activity as well).